File: native-modules/ios/MediaPipe/MediaPipePoseModule.swift
import Foundation
import UIKit
import MediaPipeTasksVision

@objc(MediaPipePose)
class MediaPipePoseModule: NSObject {

  private var poseLandmarker: PoseLandmarker?
  private var isInitialized = false

  // ============================================
  // FIRE & POLL PATTERN - Shared State
  // ============================================

  /// Ultimo risultato disponibile per il polling JS
  static var latestResult: [String: Any]?

  /// Flag per evitare elaborazioni sovrapposte
  static var isProcessing: Bool = false

  /// Timestamp dell'ultimo risultato (per evitare duplicati)
  static var lastResultTimestamp: Int64 = 0

  /// Riferimento al PoseLandmarker condiviso con il plugin
  static var sharedLandmarker: PoseLandmarker?
  static var isLandmarkerReady: Bool = false

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  /// Inizializza MediaPipe Pose Landmarker
  @objc
  func initialize(_ resolve: @escaping RCTPromiseResolveBlock,
                  rejecter reject: @escaping RCTPromiseRejectBlock) {
    do {
      if isInitialized {
        NSLog("MediaPipePose: ⚠️ Già inizializzato, skip")
        resolve(true)
        return
      }

      NSLog("MediaPipePose: 🚀 Inizializzazione MediaPipe Pose Landmarker...")

      // Debug: elenca tutti i file .task nel bundle
      NSLog("MediaPipePose: 🔍 Cercando file .task nel bundle...")
      if let bundlePath = Bundle.main.resourcePath {
        let fileManager = FileManager.default
        if let files = try? fileManager.contentsOfDirectory(atPath: bundlePath) {
          let taskFiles = files.filter { $0.hasSuffix(".task") }
          NSLog("MediaPipePose:    📂 File .task trovati: \(taskFiles)")
          if taskFiles.isEmpty {
            NSLog("MediaPipePose:    📂 Tutti i file nel bundle root: \(files.prefix(20))...")
          }
        }
      }

      // Ottieni il path del modello dal bundle
      guard let modelPath = Bundle.main.path(forResource: "pose_landmarker_full", ofType: "task") else {
        NSLog("MediaPipePose: ❌ Modello non trovato nel bundle!")
        NSLog("MediaPipePose: ❌ Bundle path: \(Bundle.main.bundlePath)")
        reject("INIT_ERROR", "Model file 'pose_landmarker_full.task' not found in bundle", nil)
        return
      }

      NSLog("MediaPipePose:    📦 Modello: \(modelPath)")

      // Configura BaseOptions con il modello
      let baseOptions = BaseOptions()
      baseOptions.modelAssetPath = modelPath

      // Opzioni bilanciate per detection affidabile (stesso setup di Android)
      let options = PoseLandmarkerOptions()
      options.baseOptions = baseOptions
      options.runningMode = .image // IMAGE mode per snapshot
      options.numPoses = 1 // Rileva max 1 persona
      options.minPoseDetectionConfidence = 0.5 // Soglia ragionevole
      options.minPosePresenceConfidence = 0.5 // Soglia ragionevole
      options.minTrackingConfidence = 0.5 // Soglia ragionevole

      NSLog("MediaPipePose:    ⚙️ RunningMode: IMAGE")
      NSLog("MediaPipePose:    ⚙️ NumPoses: 1")
      NSLog("MediaPipePose:    ⚙️ MinConfidence: 0.5")

      // Crea il landmarker
      poseLandmarker = try PoseLandmarker(options: options)
      isInitialized = true

      // Condividi il landmarker con il Frame Processor Plugin
      MediaPipePoseModule.sharedLandmarker = poseLandmarker
      MediaPipePoseModule.isLandmarkerReady = true

      NSLog("MediaPipePose: ✅ MediaPipe inizializzato con successo!")
      NSLog("MediaPipePose: ✅ Landmarker condiviso con Frame Processor Plugin")
      resolve(true)

    } catch {
      NSLog("MediaPipePose: ❌ ERRORE inizializzazione: \(error.localizedDescription)")
      reject("INIT_ERROR", "Failed to initialize MediaPipe: \(error.localizedDescription)", error)
    }
  }

  /// Rileva pose da file immagine
  @objc
  func detectPoseFromFile(_ filePath: String,
                          timestampMs: Double,
                          resolver resolve: @escaping RCTPromiseResolveBlock,
                          rejecter reject: @escaping RCTPromiseRejectBlock) {
    do {
      guard isInitialized, let poseLandmarker = poseLandmarker else {
        NSLog("MediaPipePose: ❌ Non inizializzato! Chiama initialize() prima.")
        reject("NOT_INITIALIZED", "MediaPipe not initialized. Call initialize() first.", nil)
        return
      }

      // Pulisci path
      let cleanPath = filePath.replacingOccurrences(of: "file://", with: "")
      NSLog("MediaPipePose: 📸 Caricamento immagine: \(cleanPath)")

      // Carica immagine
      guard var image = UIImage(contentsOfFile: cleanPath) else {
        NSLog("MediaPipePose: ❌ Impossibile caricare immagine da: \(cleanPath)")
        reject("DECODE_ERROR", "Failed to load image from: \(cleanPath)", nil)
        return
      }

      NSLog("MediaPipePose:    🖼️ Immagine caricata: \(image.size.width)x\(image.size.height)")

      // Ruota se necessario (camera frontale può generare orientamenti strani)
      if image.size.width > image.size.height {
        NSLog("MediaPipePose:    🔄 Rotazione (landscape → portrait)")
        image = rotateImage(image, by: 90) ?? image
        NSLog("MediaPipePose:    ✓ Dopo rotazione: \(image.size.width)x\(image.size.height)")
      }

      // Salva dimensioni originali
      let originalWidth = Int(image.size.width)
      let originalHeight = Int(image.size.height)

      // Ridimensiona per performance (MediaPipe funziona bene con 480p)
      let maxDimension: CGFloat = 480
      var processedImage = image

      if image.size.width > maxDimension || image.size.height > maxDimension {
        let scale = min(maxDimension / image.size.width, maxDimension / image.size.height)
        let newWidth = image.size.width * scale
        let newHeight = image.size.height * scale

        NSLog("MediaPipePose:    📏 Ridimensionamento: \(image.size.width)x\(image.size.height) → \(newWidth)x\(newHeight)")

        processedImage = resizeImage(image, to: CGSize(width: newWidth, height: newHeight))
      }

      NSLog("MediaPipePose:    🎯 Immagine finale per detection: \(processedImage.size.width)x\(processedImage.size.height)")

      // Converti in MPImage
      guard let mpImage = try? MPImage(uiImage: processedImage) else {
        NSLog("MediaPipePose: ❌ Impossibile creare MPImage")
        reject("IMAGE_ERROR", "Failed to create MPImage from UIImage", nil)
        return
      }

      NSLog("MediaPipePose:    🔍 Esecuzione detection...")

      // Esegui detection (IMAGE mode usa detect, non detectAsync)
      let result = try poseLandmarker.detect(image: mpImage)

      // Log risultato
      if result.landmarks.isEmpty {
        NSLog("MediaPipePose:    ⚠️ Nessuna persona rilevata!")
      } else {
        NSLog("MediaPipePose:    ✅ Persona rilevata! \(result.landmarks.count) pose, \(result.landmarks[0].count) landmarks")

        // Log primi 5 landmarks per debug
        let landmarks = result.landmarks[0]
        if landmarks.count >= 13 {
          let nose = landmarks[0]
          let leftEye = landmarks[2]
          let rightEye = landmarks[5]
          let leftShoulder = landmarks[11]
          let rightShoulder = landmarks[12]

          NSLog("MediaPipePose:    👃 Naso [0]: x=\(nose.x), y=\(nose.y), z=\(nose.z), vis=\(nose.visibility)")
          NSLog("MediaPipePose:    👁️ LeftEye [2]: x=\(leftEye.x), y=\(leftEye.y), vis=\(leftEye.visibility)")
          NSLog("MediaPipePose:    👁️ RightEye [5]: x=\(rightEye.x), y=\(rightEye.y), vis=\(rightEye.visibility)")
          NSLog("MediaPipePose:    💪 LeftShoulder [11]: x=\(leftShoulder.x), y=\(leftShoulder.y), vis=\(leftShoulder.visibility)")
          NSLog("MediaPipePose:    💪 RightShoulder [12]: x=\(rightShoulder.x), y=\(rightShoulder.y), vis=\(rightShoulder.visibility)")
          NSLog("MediaPipePose:    📐 Immagine processata: \(processedImage.size.width)x\(processedImage.size.height)")
          NSLog("MediaPipePose:    📐 Dimensioni originali: \(originalWidth)x\(originalHeight)")
        }
      }

      // Converti risultato
      let response = convertPoseLandmarkerResult(
        result,
        originalWidth: originalWidth,
        originalHeight: originalHeight,
        processedWidth: Int(processedImage.size.width),
        processedHeight: Int(processedImage.size.height)
      )

      resolve(response)

    } catch {
      NSLog("MediaPipePose: ❌ ERRORE detection: \(error.localizedDescription)")
      reject("DETECTION_ERROR", "Pose detection failed: \(error.localizedDescription)", error)
    }
  }

  /// Converte PoseLandmarkerResult nel formato React Native
  private func convertPoseLandmarkerResult(
    _ result: PoseLandmarkerResult,
    originalWidth: Int,
    originalHeight: Int,
    processedWidth: Int,
    processedHeight: Int
  ) -> [String: Any] {

    var landmarksArray: [[String: Any]] = []

    if result.landmarks.isEmpty {
      // Nessuna persona rilevata - restituisci array vuoto
      NSLog("MediaPipePose:    📊 Risposta: 0 landmarks (nessuna persona)")
      return [
        "landmarks": landmarksArray,
        "imageWidth": originalWidth,
        "imageHeight": originalHeight
      ]
    }

    // Prendi prima persona (abbiamo setNumPoses(1))
    let poseLandmarks = result.landmarks[0]

    // MediaPipe ha 33 landmarks, le coordinate sono già normalizzate 0-1
    for landmark in poseLandmarks {
      let landmarkDict: [String: Any] = [
        "x": Double(landmark.x),
        "y": Double(landmark.y),
        "z": Double(landmark.z),
        // Usa visibility come confidence (MediaPipe chiama visibility quello che ML Kit chiama inFrameLikelihood)
        "inFrameLikelihood": Double(truncating: landmark.visibility ?? 0)
      ]
      landmarksArray.append(landmarkDict)
    }

    NSLog("MediaPipePose:    📊 Risposta: \(poseLandmarks.count) landmarks (dimensioni originali: \(originalWidth)x\(originalHeight))")

    // IMPORTANTE: Restituiamo le dimensioni ORIGINALI, non quelle processate
    // Le coordinate di MediaPipe sono normalizzate e valide per qualsiasi dimensione
    return [
      "landmarks": landmarksArray,
      "imageWidth": originalWidth,
      "imageHeight": originalHeight
    ]
  }

  /// Helper: ridimensiona immagine
  private func resizeImage(_ image: UIImage, to size: CGSize) -> UIImage {
    UIGraphicsBeginImageContextWithOptions(size, false, 1.0)
    image.draw(in: CGRect(origin: .zero, size: size))
    let resizedImage = UIGraphicsGetImageFromCurrentImageContext() ?? image
    UIGraphicsEndImageContext()
    return resizedImage
  }

  /// Helper: ruota immagine
  private func rotateImage(_ image: UIImage, by degrees: CGFloat) -> UIImage? {
    let radians = degrees * .pi / 180
    var newSize = CGRect(origin: .zero, size: image.size)
      .applying(CGAffineTransform(rotationAngle: radians))
      .size

    // Trim off the extremely small float value to prevent core graphics from rounding it up
    newSize.width = floor(newSize.width)
    newSize.height = floor(newSize.height)

    UIGraphicsBeginImageContextWithOptions(newSize, false, image.scale)
    guard let context = UIGraphicsGetCurrentContext() else { return nil }

    // Move origin to middle
    context.translateBy(x: newSize.width / 2, y: newSize.height / 2)
    // Rotate around middle
    context.rotate(by: radians)
    // Draw the image at its center
    image.draw(in: CGRect(x: -image.size.width / 2, y: -image.size.height / 2,
                          width: image.size.width, height: image.size.height))

    let rotatedImage = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()

    return rotatedImage
  }

  // ============================================
  // FIRE & POLL PATTERN - Metodi per polling JS
  // ============================================

  /// Restituisce l'ultimo risultato elaborato dal Frame Processor.
  /// Chiamato in polling dal JS thread ogni ~33ms.
  @objc
  func getLastResult(_ resolve: @escaping RCTPromiseResolveBlock,
                     rejecter reject: @escaping RCTPromiseRejectBlock) {
    if let result = MediaPipePoseModule.latestResult {
      // Clona il risultato e aggiungi timestamp
      var clonedResult = result
      clonedResult["timestamp"] = Double(MediaPipePoseModule.lastResultTimestamp)
      resolve(clonedResult)
    } else {
      resolve(nil)
    }
  }

  /// Verifica se il Frame Processor è pronto
  @objc
  func isFrameProcessorReady(_ resolve: @escaping RCTPromiseResolveBlock,
                             rejecter reject: @escaping RCTPromiseRejectBlock) {
    let isReady = MediaPipePoseModule.isLandmarkerReady && MediaPipePoseModule.sharedLandmarker != nil
    resolve(isReady)
  }
}

File: native-modules/android/mediapipe/MediaPipePoseModule.kt
package {{PACKAGE_NAME}}.mediapipe

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap

class MediaPipePoseModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var poseLandmarker: PoseLandmarker? = null
    private var isInitialized = false

    // ============================================
    // FIRE & POLL PATTERN - Shared State
    // ============================================
    companion object {
        // Ultimo risultato disponibile per il polling JS
        @Volatile var latestResult: WritableMap? = null

        // Flag per evitare elaborazioni sovrapposte
        @Volatile var isProcessing: Boolean = false

        // Timestamp dell'ultimo risultato (per evitare duplicati)
        @Volatile var lastResultTimestamp: Long = 0

        // Riferimento al PoseLandmarker condiviso con il plugin
        @Volatile var sharedLandmarker: PoseLandmarker? = null
        @Volatile var isLandmarkerReady: Boolean = false

        // Riferimento al ReactContext per il plugin
        @Volatile var appReactContext: ReactApplicationContext? = null
    }

    override fun getName(): String = "MediaPipePose"

    /**
     * Inizializza MediaPipe Pose Landmarker
     */
    @ReactMethod
    fun initialize(promise: Promise) {
        try {
            if (isInitialized) {
                android.util.Log.d("MediaPipePose", "⚠️ Già inizializzato, skip")
                promise.resolve(true)
                return
            }

            android.util.Log.d("MediaPipePose", "🚀 Inizializzazione MediaPipe Pose Landmarker...")

            // Configura BaseOptions con il modello dagli assets
            // IMPORTANTE: MediaPipe richiede uno slash nel path, quindi usiamo "./"
            val baseOptions = BaseOptions.builder()
                .setModelAssetPath("./pose_landmarker_full.task")
                .build()

            android.util.Log.d("MediaPipePose", "   📦 Modello: pose_landmarker_full.task")

            // Opzioni bilanciate per detection affidabile
            val options = PoseLandmarker.PoseLandmarkerOptions.builder()
                .setBaseOptions(baseOptions)
                .setRunningMode(RunningMode.IMAGE) // IMAGE mode per snapshot
                .setNumPoses(1) // Rileva max 1 persona
                .setMinPoseDetectionConfidence(0.5f) // Soglia ragionevole
                .setMinPosePresenceConfidence(0.5f) // Soglia ragionevole
                .setMinTrackingConfidence(0.5f) // Soglia ragionevole
                .build()

            android.util.Log.d("MediaPipePose", "   ⚙️ RunningMode: IMAGE")
            android.util.Log.d("MediaPipePose", "   ⚙️ NumPoses: 1")
            android.util.Log.d("MediaPipePose", "   ⚙️ MinConfidence: 0.5")

            // Crea il landmarker
            poseLandmarker = PoseLandmarker.createFromOptions(reactContext, options)
            isInitialized = true

            // Condividi il landmarker e context con il Frame Processor Plugin
            sharedLandmarker = poseLandmarker
            isLandmarkerReady = true
            appReactContext = reactContext

            android.util.Log.d("MediaPipePose", "✅ MediaPipe inizializzato con successo!")
            android.util.Log.d("MediaPipePose", "✅ Landmarker condiviso con Frame Processor Plugin")
            promise.resolve(true)

        } catch (e: Exception) {
            android.util.Log.e("MediaPipePose", "❌ ERRORE inizializzazione: ${e.message}")
            e.printStackTrace()
            promise.reject("INIT_ERROR", "Failed to initialize MediaPipe: ${e.message}", e)
        }
    }

    /**
     * Rileva pose da file immagine
     */
    @ReactMethod
    fun detectPoseFromFile(filePath: String, timestampMs: Double, promise: Promise) {
        try {
            if (!isInitialized || poseLandmarker == null) {
                android.util.Log.e("MediaPipePose", "❌ Non inizializzato! Chiama initialize() prima.")
                promise.reject("NOT_INITIALIZED", "MediaPipe not initialized. Call initialize() first.")
                return
            }

            // Pulisci path
            val cleanPath = filePath.replace("file://", "")
            android.util.Log.d("MediaPipePose", "📸 Caricamento immagine: $cleanPath")

            // Carica bitmap
            var bitmap = BitmapFactory.decodeFile(cleanPath)
            if (bitmap == null) {
                android.util.Log.e("MediaPipePose", "❌ Impossibile caricare bitmap da: $cleanPath")
                promise.reject("DECODE_ERROR", "Failed to load image from: $cleanPath")
                return
            }

            android.util.Log.d("MediaPipePose", "   🖼️ Bitmap caricato: ${bitmap.width}x${bitmap.height}")

            // Ruota se landscape (camera frontale portrait genera landscape)
            if (bitmap.width > bitmap.height) {
                android.util.Log.d("MediaPipePose", "   🔄 Rotazione 90° (landscape → portrait)")
                val matrix = Matrix()
                matrix.postRotate(90f)
                val rotatedBitmap = Bitmap.createBitmap(
                    bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true
                )
                bitmap.recycle()
                bitmap = rotatedBitmap
                android.util.Log.d("MediaPipePose", "   ✓ Dopo rotazione: ${bitmap.width}x${bitmap.height}")
            }

            // Salva dimensioni originali
            val originalWidth = bitmap.width
            val originalHeight = bitmap.height

            // Ridimensiona per performance (MediaPipe funziona bene con 480p)
            val maxDimension = 480
            var processedBitmap = bitmap

            if (bitmap.width > maxDimension || bitmap.height > maxDimension) {
                val scale = minOf(
                    maxDimension.toFloat() / bitmap.width,
                    maxDimension.toFloat() / bitmap.height
                )

                val newWidth = (bitmap.width * scale).toInt()
                val newHeight = (bitmap.height * scale).toInt()

                android.util.Log.d("MediaPipePose", "   📏 Ridimensionamento: ${bitmap.width}x${bitmap.height} → ${newWidth}x${newHeight}")

                processedBitmap = Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
                if (processedBitmap != bitmap) {
                    bitmap.recycle()
                }
            }

            android.util.Log.d("MediaPipePose", "   🎯 Immagine finale per detection: ${processedBitmap.width}x${processedBitmap.height}")

            // Converti in MPImage
            val mpImage = BitmapImageBuilder(processedBitmap).build()

            android.util.Log.d("MediaPipePose", "   🔍 Esecuzione detection...")

            // Esegui detection (IMAGE mode usa detect, non detectForVideo)
            val result = poseLandmarker!!.detect(mpImage)

            // Log risultato
            if (result.landmarks().isEmpty()) {
                android.util.Log.w("MediaPipePose", "   ⚠️ Nessuna persona rilevata!")
            } else {
                android.util.Log.d("MediaPipePose", "   ✅ Persona rilevata! ${result.landmarks().size} pose, ${result.landmarks()[0].size} landmarks")

                // Log primi 5 landmarks per debug
                val nose = result.landmarks()[0][0]
                val leftEye = result.landmarks()[0][2]
                val rightEye = result.landmarks()[0][5]
                val leftShoulder = result.landmarks()[0][11]
                val rightShoulder = result.landmarks()[0][12]

                android.util.Log.d("MediaPipePose", "   👃 Naso [0]: x=${nose.x()}, y=${nose.y()}, z=${nose.z()}, vis=${nose.visibility().orElse(0f)}")
                android.util.Log.d("MediaPipePose", "   👁️ LeftEye [2]: x=${leftEye.x()}, y=${leftEye.y()}, vis=${leftEye.visibility().orElse(0f)}")
                android.util.Log.d("MediaPipePose", "   👁️ RightEye [5]: x=${rightEye.x()}, y=${rightEye.y()}, vis=${rightEye.visibility().orElse(0f)}")
                android.util.Log.d("MediaPipePose", "   💪 LeftShoulder [11]: x=${leftShoulder.x()}, y=${leftShoulder.y()}, vis=${leftShoulder.visibility().orElse(0f)}")
                android.util.Log.d("MediaPipePose", "   💪 RightShoulder [12]: x=${rightShoulder.x()}, y=${rightShoulder.y()}, vis=${rightShoulder.visibility().orElse(0f)}")
                android.util.Log.d("MediaPipePose", "   📐 Bitmap processato: ${processedBitmap.width}x${processedBitmap.height}")
                android.util.Log.d("MediaPipePose", "   📐 Dimensioni originali: ${originalWidth}x${originalHeight}")
            }

            // Converti risultato
            val response = convertPoseLandmarkerResult(
                result,
                originalWidth,
                originalHeight,
                processedBitmap.width,
                processedBitmap.height
            )

            promise.resolve(response)
            processedBitmap.recycle()

        } catch (e: Exception) {
            android.util.Log.e("MediaPipePose", "❌ ERRORE detection: ${e.message}")
            e.printStackTrace()
            promise.reject("DETECTION_ERROR", "Pose detection failed: ${e.message}", e)
        }
    }

    /**
     * Converte PoseLandmarkerResult nel formato React Native
     */
    private fun convertPoseLandmarkerResult(
        result: PoseLandmarkerResult,
        originalWidth: Int,
        originalHeight: Int,
        processedWidth: Int,
        processedHeight: Int
    ): WritableNativeMap {
        val response = WritableNativeMap()
        val landmarksArray = WritableNativeArray()

        if (result.landmarks().isEmpty()) {
            // Nessuna persona rilevata - restituisci array vuoto
            response.putArray("landmarks", landmarksArray)
            response.putInt("imageWidth", originalWidth)
            response.putInt("imageHeight", originalHeight)
            android.util.Log.d("MediaPipePose", "   📊 Risposta: 0 landmarks (nessuna persona)")
            return response
        }

        // Prendi prima persona (abbiamo setNumPoses(1))
        val poseLandmarks = result.landmarks()[0]

        // MediaPipe ha 33 landmarks, le coordinate sono già normalizzate 0-1
        for (landmark in poseLandmarks) {
            val landmarkMap = WritableNativeMap()

            // Coordinate già normalizzate da MediaPipe
            landmarkMap.putDouble("x", landmark.x().toDouble())
            landmarkMap.putDouble("y", landmark.y().toDouble())
            landmarkMap.putDouble("z", landmark.z().toDouble())

            // Usa visibility come confidence (MediaPipe chiama visibility quello che ML Kit chiama inFrameLikelihood)
            val confidence = landmark.visibility().orElse(0.5f).toDouble()
            landmarkMap.putDouble("inFrameLikelihood", confidence)

            landmarksArray.pushMap(landmarkMap)
        }

        response.putArray("landmarks", landmarksArray)
        // ⚠️ IMPORTANTE: Restituiamo le dimensioni ORIGINALI, non quelle processate
        // Le coordinate di MediaPipe sono normalizzate e valide per qualsiasi dimensione
        response.putInt("imageWidth", originalWidth)
        response.putInt("imageHeight", originalHeight)

        android.util.Log.d("MediaPipePose", "   📊 Risposta: ${poseLandmarks.size} landmarks (dimensioni originali: ${originalWidth}x${originalHeight})")

        return response
    }

    // ============================================
    // FIRE & POLL PATTERN - Metodo per polling JS
    // ============================================

    /**
     * Restituisce l'ultimo risultato elaborato dal Frame Processor.
     * Chiamato in polling dal JS thread ogni ~33ms.
     */
    @ReactMethod
    fun getLastResult(promise: Promise) {
        try {
            val result = latestResult
            if (result != null) {
                // Clona il risultato per evitare race condition
                val clonedResult = Arguments.createMap()
                clonedResult.merge(result)
                clonedResult.putDouble("timestamp", lastResultTimestamp.toDouble())
                promise.resolve(clonedResult)
            } else {
                promise.resolve(null)
            }
        } catch (e: Exception) {
            android.util.Log.e("MediaPipePose", "❌ Errore getLastResult: ${e.message}")
            promise.resolve(null)
        }
    }

    /**
     * Verifica se il Frame Processor è pronto
     */
    @ReactMethod
    fun isFrameProcessorReady(promise: Promise) {
        promise.resolve(isLandmarkerReady && sharedLandmarker != null)
    }

    /**
     * Cleanup
     */
    override fun invalidate() {
        super.invalidate()
        poseLandmarker?.close()
        poseLandmarker = null
        isInitialized = false

        // Pulisci anche lo stato condiviso
        sharedLandmarker = null
        isLandmarkerReady = false
        latestResult = null
        appReactContext = null

        android.util.Log.d("MediaPipePose", "🛑 MediaPipe chiuso")
    }
}